sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteCollationJoin.scala

```diff
@@ -0,0 +1,102 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.analysis
+
+import org.apache.spark.sql.catalyst.expressions.{AttributeReference, CollationKey, EqualNullSafe, EqualTo, Lower}
+import org.apache.spark.sql.catalyst.plans.logical.{Join, LogicalPlan}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.catalyst.util.CollationFactory
+import org.apache.spark.sql.types.{ArrayType, StringType, StructType}
+
+object RewriteCollationJoin extends Rule[LogicalPlan] {
+  def apply(plan: LogicalPlan): LogicalPlan = plan transformUpWithNewOutput {
+    case j @ Join(_, _, _, Some(condition), _) =>
+      val newCondition = condition transform {
+        case EqualTo(l: AttributeReference, r: AttributeReference) =>
+          (l.dataType, r.dataType) match {
+            case (_: StringType, _: StringType) =>
+              val collationId = l.dataType.asInstanceOf[StringType].collationId
+              val collation = CollationFactory.fetchCollation(collationId)
+              if (collation.supportsBinaryEquality) {
+                return plan
```

Comment:
no, we shouldn't `return` within the transform lambda. The plan may have multiple joins and the `return` here reverts all rewrites.

Reply:
Isn't `plan` here only at the "Join" level, i.e. this should return the original "Join" as it was, without modification

if there is another join, it should be a completely separate visit to this rule (which is why now no tests are failing with multiple Joins, such as numerous TPCDS queries with collation)